edb/edgeql-parser/edgeql-parser-python/src/parser.rs

```diff
@@ -0,0 +1,163 @@
+use std::collections::HashMap;
+
+use cpython::{
+    ObjectProtocol, PyClone, PyInt, PyList, PyObject, PyResult, PyString, PyTuple, Python,
+    PythonObject, PythonObjectWithCheckedDowncast, ToPyObject,
+};
+
+use edgeql_parser::parser;
+
+use crate::errors::{parser_error_into_tuple, ParserResult};
+use crate::pynormalize::value_to_py_object;
+use crate::tokenizer::OpaqueToken;
+
+pub fn parse(py: Python, parser_name: &PyString, tokens: PyObject) -> PyResult<PyTuple> {
+    let (spec, productions) = load_spec(py, parser_name.to_string(py)?.as_ref())?;
+
+    let tokens = downcast_tokens(py, tokens)?;
+
+    let context = parser::Context::new(spec);
+    let (cst, errors) = parser::parse(&tokens, &context);
+
+    let cst = cst.map(|c| to_py_cst(c, py)).transpose()?;
+
+    let errors = errors
+        .into_iter()
+        .map(|e| parser_error_into_tuple(py, e))
+        .collect::<Vec<_>>();
+    let errors = PyList::new(py, &errors);
+
+    let res = ParserResult::create_instance(py, cst.into_py_object(py), errors)?;
+
+    Ok((res, productions).into_py_object(py))
+}
+
+py_class!(pub class CSTNode |py| {
+    data _production: PyObject;
+    data _terminal: PyObject;
+
+    def production(&self) -> PyResult<PyObject> {
+        Ok(self._production(py).clone_ref(py))
+    }
+    def terminal(&self) -> PyResult<PyObject> {
+        Ok(self._terminal(py).clone_ref(py))
+    }
+});
+
+py_class!(pub class Production |py| {
+    data _id: PyInt;
+    data _args: PyList;
+
+    def id(&self) -> PyResult<PyInt> {
+        Ok(self._id(py).clone_ref(py))
+    }
+    def args(&self) -> PyResult<PyList> {
+        Ok(self._args(py).clone_ref(py))
+    }
+});
+
+py_class!(pub class Terminal |py| {
+    data _text: PyString;
+    data _value: PyObject;
+    data _start: u64;
+    data _end: u64;
+
+    def text(&self) -> PyResult<PyString> {
+        Ok(self._text(py).clone_ref(py))
+    }
+    def value(&self) -> PyResult<PyObject> {
+        Ok(self._value(py).clone_ref(py))
+    }
+    def start(&self) -> PyResult<u64> {
+        Ok(*self._start(py))
+    }
+    def end(&self) -> PyResult<u64> {
+        Ok(*self._end(py))
+    }
+});
+
+static mut PARSER_SPECS: Option<HashMap<String, (parser::Spec, PyObject)>> = None;
```

Comment:
Actually, could we use `lazy_static` and a `Mutex`/`RwLock` instead of doing unsafe stuff? (Here and in the other place we do a trick like this.)

It's cleaner in general and it'll matter if python ever drops the GIL :P